<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>GraftM - How to get fast community profiles from metagenomes</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="">

    <!-- Le HTML5 shim, for IE6-8 support of HTML elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- Le styles -->
    <link href="./css/bootstrap.min.css" rel="stylesheet">
    <link href="./css/bootstrap-responsive.min.css" rel="stylesheet">
    <link href="./css/docs.css" rel="stylesheet">
  </head>

  <body class="preview" data-spy="scroll" data-target=".subnav" data-offset="80">
    <script src="./js/bsa.js"></script>

    <div class="container">

<br>
<header class="jumbotron subhead" id="overview">
  <div class="row">
    <img style="display: block; margin-left: auto; margin-right: auto;" src="./img/graftm_logo_sm.png"/>
  </div>
</header>


<section id="theboss">
  <!-- Headings & Paragraph Copy -->
  <hr>
  <h1>Overview</h1>
  <p>GraftM is a meta-omic tool that identifies and classifies marker genes in short read datasets (metagenomes and metatranscriptomes), as well as assembled contigs, whole genomes and protein sequences. GraftM outputs a taxonomic/functional summary table, a <a href="http://sourceforge.net/p/krona/home/krona/">krona</a> plot, as well as various other run statistics. Both unaligned and aligned "hit" sequences are provided.
  <p>GraftM is designed for speed and accuracy. It is able to find marker genes in a 200Mb of assembled metagenome in &lt;20 sec, and compares favourably with similar tools in accuracy benchmarking.</p>
  <hr>
  <h1>News + Updates</h1>
  <p>31/03/2015 - Please submit any requests for features in GraftM to the GitHub page!</p>
  <hr>
  <div class="row">
    <div class="span4">
        <h1>Getting GraftM</h1>
        <p>GraftM has a few dependencies, which are needed to use all the features of GraftM. The <a href="https://github.com/geronimp/graftM/wiki">manual</a> has more detailed instructions on how to install GraftM, its dependencies, and a description of command line options.</p>
        <p>The source code for GraftM is also available on <a href="https://github.com/geronimp/graftM.git">GitHub</a>.</p>
        <hr>
        <h1>Licensing</h1>
        <p>Written by <a href="http://ecogenomic.org/users/joel-boyd">Joel Boyd</a>, <a href="http://ecogenomic.org/users/ben-woodcroft">Ben Woodcroft</a> and <a href="http://ecogenomic.org/users/gene-tyson">Gene Tyson</a>, GraftM is licensed using the Lesser GNU General Public License version 3 as published by the Free Software Foundation.</p>

    </div>
    <div class="span4">
        <h1>Using GraftM</h1>
        <p>GraftM was developed as a tool for the rapid analysis of short read metagenomic data. It was developed as a tool that could search and classify marker gene reads, without using pairwise sequence comparisons such as BLAST. The result is a flexible, easy-to-use tool that can be run on any gene: all in a fraction of the time required by other methods.</p>
        <p>We provide some GraftM packages, such as 16S rRNA, mcrA and pmoA, but for detailed instructions on how to create your own GraftM package, please visit the manual page.</p>
        <hr>
        <h1>Example Usage</h1>
        <p>Here is just a quick example of how to run GraftM. You can find some example data to use GraftM with <a href="https://github.com/geronimp/graftM/tree/master/example_data">here</a>.</p>
        <p><code class="codeblock">$ graftM graft --forward reads_f.fa --graftm_package mcrA_gpkg</code></p>
    </div>
    <div class="span4">
        <h1>Citing GraftM</h1>
        <p>If you use GraftM then please cite us. As of yet we have no publication, but please cite the GitHub page.</p>
        <hr>
        <h1>Contact</h1>
        <p>Of course, you can always contact <a href="https://github.com/geronimp">Joel</a> or <a href="https://github.com/wwood">Ben</a> on the <a href="https://github.com/geronimp/graftM/issues">issues page</a> for any issues that you encounter with GraftM. Alternatively you could just send us an email through supportm@googlegroups.com. We're more than happy to help.</p>
        <hr>
        <h1>The M suite</h1>
        <p>GraftM is part of the M(etagenomics) suite of tools being developed at the Australian Centre for Ecogenomics (<a href="http://ecogenomic.org/">ACE</a>). We'd love for you to check out our tools:</p>
        <ul>
            <li><strong><a href="http://minillinim.github.io/GroopM/">GroopM</a></strong> - Recover (bin) population genomes from assembled metagenomes using differential coverage across multiple datasets.</li>
            <li><strong><a href="http://ecogenomics.github.io/CheckM/">CheckM</a></strong> - Assess the quality of genomes recovered from isolates, single cells or metagenomes using lineage specific marker sets.</li>
            <li><strong><a href="http://minillinim.github.io/BamM/">BamM</a></strong> - Make and parse metagenomic BAM files.</li>
        </ul>
    </div>
  </div>
<hr>
<p>This website's structure and the GraftM logo are copyright Mike Imelfort. This site was created using a template created by the wonderful people at <a href="http://bootswatch.com">bootswatch</a>.</p>
</section>

    <!-- Placed at the end of the document so the pages load faster -->
    <script src="http://ajax.googleapis.com/ajax/libs/jquery/1.7.2/jquery.min.js"></script>
    <script src="./js/bootstrap.min.js"></script>
  </body>
</html>

