# GraftM

---

## Overview
Basically, GraftM is a pipeline used for identifying and classifying marker gene reads from metagenomic datasets. If you are after a tool that provide rapid and accurate analysis of short read sequence data, then you'll probably want to give GraftM a try

---

## Installation
Currently, GraftM has a few depedancies making installation a bit more difficult that simply using pip. We're currently working on a way to make installation of GraftM this as smooth as possible.
For now though, installation will have to be manual but very soon we hope to make this easier.
### Dependencies
* biopython v. >= 1.64 (http://biopython.org/wiki/Main_Page)
* orfm v. >= 0.1.1 (https://github.com/wwood/OrfM)
* hmmer v. >= 3.1b1 (http://hmmer.janelia.org/)
* fxtract (https://github.com/ctSkennerton/fxtract)
* pplacer v. >= 2.6.32 (http://matsen.fhcrc.org/pplacer/)
* seqmagick v. >= 0.6.0 (https://github.com/fhcrc/seqmagick)
* sra toolkit v. >= 2.3.2-5 (https://github.com/ncbi/sratoolkit)
* pandas (https://github.com/pydata/pandas)
* krona v. >= 2.4 (http://sourceforge.net/p/krona/home/krona/)

---

## Help + Contact
Of course, can always contact Joel at joel.boyd@uqconnect.edu.au for any queries, complaints or reccomendations you have about GraftM. I'm happy to help.
A manual page and pdf will be available shortly
